src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/TempPE/DesignTimeInputsCompiler.cs

```diff
@@ -0,0 +1,323 @@
+        /// </summary>
+        internal bool CompileSynchronously { get; set; }
+
+        protected override Task InitializeCoreAsync(CancellationToken cancellationToken)
+        {
+            // Create an action block process the design time delta
+            _deltaActionBlock = DataflowBlockSlim.CreateActionBlock<IProjectVersionedValue<DesignTimeInputsDelta>>(ProcessDataflowChanges);
+
+            _changeTrackerLink = _changeTracker.SourceBlock.LinkTo(_deltaActionBlock, DataflowOption.PropagateCompletion);
+
+            return Task.CompletedTask;
+        }
+
+        internal void ProcessDataflowChanges(IProjectVersionedValue<DesignTimeInputsDelta> obj)
+        {
+            // Cancel any in-progress queue processing
+            _compilationCancellationSource?.Cancel();
+
+            // Capture the latest state
+            _state = obj.Value;
+
+            // add all of the changes to our queue
+            foreach (DesignTimeInputFileChange item in _state.ChangedInputs)
+            {
+                ImmutableInterlocked.TryAdd(ref _filesToCompile, item.File, item.IgnoreFileWriteTime);
+            }
+
+            // remove any items we have queued that aren't in the inputs any more
+            foreach ((string file, _) in _filesToCompile)
+            {
+                if (!_state.AllInputs.Contains(file))
+                {
+                    ImmutableInterlocked.TryRemove(ref _filesToCompile, file, out _);
+                }
+            }
+
+            // Create a cancellation source so we can cancel the compilation if another message comes through
+            _compilationCancellationSource = CancellationTokenSource.CreateLinkedTokenSource(_project.Services.ProjectAsynchronousTasks.UnloadCancellationToken);
+
+            JoinableTask task = _scheduler.ScheduleAsyncTask(ProcessCompileQueueAsync, _compilationCancellationSource.Token);
+            
+            // For unit testing purposes, optionally block the thread until the task we scheduled is complete
+            if (CompileSynchronously)
+            {
+                _threadingService.ExecuteSynchronously(() => task.Task);
+            }
+        }
+
+        protected override async Task DisposeCoreAsync(bool initialized)
+        {
+            if (_deltaActionBlock != null)
+            {
+                // This will stop our blocks taking any more input
+                _deltaActionBlock.Complete();
+
+                await _deltaActionBlock.Completion;
+            }
+
+            _compilationCancellationSource?.Dispose();
+            _scheduler.Dispose();
+            _changeTrackerLink?.Dispose();
+        }
+
+        private Task ProcessCompileQueueAsync(CancellationToken token)
+        {
+            int compileCount = 0;
+            int initialQueueLength = _filesToCompile.Count;
+            var compileStopWatch = Stopwatch.StartNew();
+            return _activeWorkspaceProjectContextHost.OpenContextForWriteAsync(async accessor =>
```

Comment:
I looked briefly and it seems possible to update `OpenContextForWriteAsync` to accept a cancellation token. It's not clear to me whether that'd be important in this scenario or not. The telemetry here might reveal that eventually.

It's possible that there are other consumers of this and related APIs that would benefit from cancellation support on what I presume to be a widely-scoped lock.

Reply:
Definitely something for another PR, and something I'm not qualified to comment on. Any thoughts @davkean?

Reply:
Yeah there could be benefit here. I think I avoided it at the time because it ballooned overloads and we didn't need it.